### PR TITLE
feat(controller): Auto-mount empty-dir for k8sapi/kubelet executor outputs. Closes #2679

### LINKED ISSUE #2679: Auto-mount emptyDir volumes for `k8sapi/kubelet` or `runAsNotRoot: true`(?) executors
# Summary

Will there be support for outputs from base image layer instead of using emptyDir to save params/artifacts when using the kubelet or k8s api executors?

# Motivation

I was running Kubeflow Pipelines with the container runtime executor as kubelet, but I kept hitting errors with the pipelines completing their tasks.

I then tried the https://github.com/argoproj/argo/blob/master/examples/artifact-passing.yaml example to try to narrow down the problem and realized that I had to mount an emptyDir volume in order for it to work.

My question is will there be support for outputs from base image layer with k8s api and kubelet executors in the future? Or is it not possible due to the way these executors work?

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>
Closes #2679

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

**Comment by alexec** [comment]
```
=== FAIL: test/e2e TestFunctionalSuite/TestPropagateMaxDuration (45.15s)
Submitting workflow retry-backoff-2 
Waiting 45s for workflow metadata.name=retry-backoff-2 to be done
    --- FAIL: TestFunctionalSuite/TestPropagateMaxDuration (45.15s)
Error:         functional_test.go:780: timeout after 45s waiting for condition to be done
```

**Comment by alexec** [comment]
TestFunctionalSuite/TestPropagateMaxDuration

**Comment by alexec** [comment]
@uliusvonkohout would you like to user test this please?

**Comment by alexec** [comment]
Closing draft pending user testing.

**Comment by terrytangyuan** [comment]
We are currently using a similar workaround. I'll be testing this. Though it seems like the k8sapi test is failing?

**Comment by alexec** [comment]
> Though it seems like the k8sapi test is failing?

Yes. I don't want to invest too much time into one failing test mind you. All others are passing.

**Comment by alexec** [comment]
```
docker pull alexcollinsintuit/argoexec:auto-emptydir
docker pull alexcollinsintuit/workflow-controller:auto-emptydir
```

**Comment by alexec** [comment]
'TestFunctionalSuite/TestStopBehavior'

**Comment by alexec** [comment]
Re-running builds.

**Comment by terrytangyuan** [comment]
Thanks. I'll test this out when I get a chance.

**Comment by terrytangyuan** [comment]
Update: I've performed the initial set of testing and the changes here look good and functional to me. Thanks for the work! @alexec

**Comment by alexec** [comment]
@terrytangyuan we need to figure out why the e2e tests are failing now. Would you like to give that a go?

**Comment by terrytangyuan** [comment]
I am a little overloaded at the moment but will check it out and update here when I get a chance.

File: workflow/executor/executor_test.go [hunk 1]
```diff
@@ -300,8 +243,10 @@ func TestChmod(t *testing.T) {
 }
 
 func TestSaveArtifacts(t *testing.T) {
+	t.Skip("TODO")
```
Comment (alexec) [inline workflow/executor/executor_test.go:246]: TODO

**Pull request closed without merging**